cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/dataset/partitioned/DynamicPartitioningOutputFormat.java

```diff
@@ -0,0 +1,189 @@
+ * Dataset.
+ *
+ * @param <K> Type of key
+ * @param <V> Type of value
+ */
+public class DynamicPartitioningOutputFormat<K, V> extends FileOutputFormat<K, V> {
+
+  public static final String HCONF_ATTR_OUTPUT_DATASET = "output.dataset.name";
+
+  // TODO: keep the limit to only FileOutputFormat and FileOutputCommitter?
+  private FileOutputFormat<K, V> fileOutputFormat;
+  private FileOutputCommitter committer;
+
+  /**
+   * Create a composite record writer that can write key/value data to different output files.
+   *
+   * @return a composite record writer
+   * @throws IOException
+   */
+  @Override
+  public RecordWriter<K, V> getRecordWriter(final TaskAttemptContext job) throws IOException {
+    final String outputName = FileOutputFormat.getOutputName(job);
+
+    Class<? extends DynamicPartitioner> className = job.getConfiguration()
+      .getClass(PartitionedFileSetArguments.DYNAMIC_PARTITIONER_CLASS_NAME, null, DynamicPartitioner.class);
+
+    @SuppressWarnings("unchecked")
+    final DynamicPartitioner<K, V> dynamicPartitioner =
+      new InstantiatorFactory(false).get(TypeToken.of(className)).create();
+
+    MapReduceMetrics.TaskType taskType = MapReduceMetrics.TaskType.from(job.getTaskAttemptID().getTaskType());
+    final BasicMapReduceTaskContext<K, V> mrTaskContext = new MapReduceTaskContextProvider(job, taskType).get();
+    String outputDatasetName = job.getConfiguration().get(HCONF_ATTR_OUTPUT_DATASET);
+    PartitionedFileSet outputDataset = mrTaskContext.getDataset(outputDatasetName);
+    final Partitioning partitioning = outputDataset.getPartitioning();
+
+    dynamicPartitioner.initialize(mrTaskContext);
+
+    return new RecordWriter<K, V>() {
+
+      // a cache storing the record writers for different output files.
+      TreeMap<String, RecordWriter<K, V>> recordWriters = new TreeMap<>();
+
+      public void write(K key, V value) throws IOException, InterruptedException {
+        PartitionKey partitionKey = dynamicPartitioner.getPartitionKey(key, value);
+        String relativePath = PartitionedFileSetDataset.getOutputPath(partitionKey, partitioning);
+        String finalPath = relativePath + "/" + outputName;
+
+        RecordWriter<K, V> rw = this.recordWriters.get(finalPath);
+        if (rw == null) {
+          // if we don't have the record writer yet for the final path, create one and add it to the cache
+          rw = getBaseRecordWriter(getTaskAttemptContext(job, finalPath));
+          this.recordWriters.put(finalPath, rw);
+        }
+        rw.write(key, value);
+      }
+
+      @Override
+      public void close(TaskAttemptContext context) throws IOException, InterruptedException {
+        try {
+          for (RecordWriter<K, V> recordWriter : this.recordWriters.values()) {
+            recordWriter.close(context);
+          }
+          this.recordWriters.clear();
+
+          mrTaskContext.flushOperations();
```

Comment:
We create the mrTaskContext and pass it to the dynamicPartitioner on line 86, so we close the mrTaskContext here.